File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int N = nums.length;
        if (N == 1) return nums[0];
        return Math.max(helper(nums, 0, N - 1), helper(nums, 1, N));
    }

    public int helper(int[] nums, int start, int stop) {
        int current = 0, next = 0;
        for (int i = start; i < stop; ++i) {
            int max = Math.max(current + nums[i], next);
            current = next;
            next = max;
        }

        return next;
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]

        def robber(houses):
            m = len(houses)

            def helper(i):
                if i < 0:
                    return 0

                if i in memo:
                    return memo[i]
                
                memo[i] = max(helper(i - 1), helper(i - 2) + houses[i])
                return memo[i]

            memo = {}
            return helper(m - 1)
  
        return max(robber(nums[:-1]), robber(nums[1:]))